platform-sdk/swirlds-common/src/main/java/com/swirlds/common/wiring/model/internal/deterministic/DeterministicTaskScheduler.java

```diff
@@ -0,0 +1,124 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.common.wiring.model.internal.deterministic;
+
+import com.swirlds.common.wiring.counters.ObjectCounter;
+import com.swirlds.common.wiring.model.TraceableWiringModel;
+import com.swirlds.common.wiring.schedulers.TaskScheduler;
+import com.swirlds.common.wiring.schedulers.builders.TaskSchedulerType;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+/**
+ * A deterministic implementation of a task scheduler.
+ *
+ * @param <OUT> the output type of the scheduler (use {@link Void} for a task scheduler with no output type)
+ */
+public class DeterministicTaskScheduler<OUT> extends TaskScheduler<OUT> {
+
+    private final Consumer<Runnable> submitWork;
+
+    private final ObjectCounter onRamp;
+    private final ObjectCounter offRamp;
+
+    /**
+     * Constructor.
+     *
+     * @param model               the wiring model containing this task scheduler
+     * @param name                the name of the task scheduler
+     * @param type                the type of task scheduler
+     * @param onRamp              counts when things are added to this scheduler to be eventually handled
+     * @param offRamp             counts when things are handled by this scheduler
+     * @param flushEnabled        if true, then {@link #flush()} will be enabled, otherwise it will throw.
+     * @param squelchingEnabled   if true, then squelching will be enabled, otherwise trying to squelch will throw.
+     * @param insertionIsBlocking when data is inserted into this task scheduler, will it block until capacity is
+     *                            available?
+     * @param submitWork          a method where all work should be submitted
+     */
+    protected DeterministicTaskScheduler(
+            @NonNull final TraceableWiringModel model,
+            @NonNull final String name,
+            @NonNull final TaskSchedulerType type,
+            @NonNull final ObjectCounter onRamp,
+            @NonNull final ObjectCounter offRamp,
+            final boolean flushEnabled,
+            final boolean squelchingEnabled,
+            final boolean insertionIsBlocking,
+            @NonNull final Consumer<Runnable> submitWork) {
+        super(model, name, type, flushEnabled, squelchingEnabled, insertionIsBlocking);
+
+        this.onRamp = Objects.requireNonNull(onRamp);
+        this.offRamp = Objects.requireNonNull(offRamp);
+        this.submitWork = Objects.requireNonNull(submitWork);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public long getUnprocessedTaskCount() {
+        return onRamp.getCount();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void flush() {
+        throw new UnsupportedOperationException("Flush is not supported on deterministic task schedulers");
```

Comment:
What happens to the run-of-the-mill flush requests that exist in the system, when all schedulers are deterministic?

Reply:
This is currently an unsolved problem from my perspective. Mostly likely what we will need to do is to mock out flushing operations during turtle tests (at least for turtle v1).